###################################

rm(list=ls(all=TRUE))

############ Syn mut
unzip("../../Body/3Results/AllGenesCodonUsageNoOverlap.txt.zip", exdir = "../../Body/3Results/")
SynNuc = read.table("../../Body/3Results/AllGenesCodonUsageNoOverlap.txt", header = TRUE, sep = '\t')
if (file.exists("../../Body/3Results/AllGenesCodonUsageNoOverlap.txt")) file.remove("../../Body/3Results/AllGenesCodonUsageNoOverlap.txt")

# SynNuc = read.table("../../Body/3Results/AllGenesCodonUsageNoOverlap.txt", header = TRUE)
names(SynNuc)

### make ND6 complementary:
NotND6 = SynNuc[SynNuc$Gene != 'ND6',]
ND6 = SynNuc[SynNuc$Gene == 'ND6',]
A = ND6$NeutralT
T = ND6$NeutralA
G = ND6$NeutralC
C = ND6$NeutralG
ND6$NeutralA = A
ND6$NeutralT = T
ND6$NeutralG = G
ND6$NeutralC = C
SynNuc = rbind(NotND6,ND6)

### count fraction of nucleotides
SynNuc$FrA = SynNuc$NeutralA / (SynNuc$NeutralA + SynNuc$NeutralT + SynNuc$NeutralG + SynNuc$NeutralC)
SynNuc$FrT = SynNuc$NeutralT / (SynNuc$NeutralA + SynNuc$NeutralT + SynNuc$NeutralG + SynNuc$NeutralC) 
SynNuc$FrG = SynNuc$NeutralG / (SynNuc$NeutralA + SynNuc$NeutralT + SynNuc$NeutralG + SynNuc$NeutralC) 
SynNuc$FrC = SynNuc$NeutralC / (SynNuc$NeutralA + SynNuc$NeutralT + SynNuc$NeutralG + SynNuc$NeutralC) 

SynNuc$TAXON = SynNuc$Class
SynNucAll = SynNuc
VecOfTaxa = unique(SynNuc$TAXON)
VecOfTaxaShort = c('Actinopterygii','Reptilia','Aves','Mammalia','Amphibia')

############# Longevity 
AA = read.table("../../Body/1Raw/anage_data.txt", header = TRUE, sep = '\t')
AA$Species = paste(AA$Genus,AA$Species,sep = '_')

############# data for PICs

library(ape)

tree <- read.tree("../../Body/1Raw/mtalign.aln.treefile.rooted")
data = AGG[which(as.character(AGG$Species) %in% tree$tip.label),]

df_vec <- as.character(AGG$Species)
tree_vec <- tree$tip.label

a <- setdiff(df_vec, tree_vec)
b <- setdiff(tree_vec, df_vec)

data = data[-597,]
row.names(data) = data$Species

tree2 <- drop.tip(tree, b)

TempData = data[, -1]
contrasts <- as.data.frame(apply(TempData, 2, pic, tree2))
names(contrasts) = names(TempData)


pdf("../../Body/4Figures/WholeGenomeAnalyses.EcologyAndMutSpecChordataNoOverlap.R.01.pdf", width = 25, height = 25)
###########
for (i in 1:length(VecOfTaxaShort))
{ # i = 3  = aves; i = 2 -> reptilia; i = 5 -> amphibia; i = 1 => actionopterygii
  VecOfTaxaShort[i]
  SynNuc = SynNucAll
  SynNuc = SynNuc[SynNuc$TAXON == VecOfTaxaShort[i],]

  ############ merge with ecology
  SynNucAA = merge(AA,SynNuc)
  length(unique(SynNucAA$Species))  # 192 species

  ########### question 1: which nucleotides better correlate with GT
  # AGG = aggregate(list(SynNucAA$FrA,SynNucAA$FrT,SynNucAA$FrG,SynNucAA$FrC), by = list(SynNucAA$Species,SynNucAA$Female.maturity..days.), FUN = mean)
  AGG = aggregate(list(SynNucAA$FrA,SynNucAA$FrT,SynNucAA$FrG,SynNucAA$FrC), by = list(SynNucAA$Species,SynNucAA$Maximum.longevity..yrs.), FUN = mean)
  names(AGG) = c('Species','FemaleMaturityDays','FrA','FrT','FrG','FrC')
  nrow(AGG)
  
  ###### Add PICs
  data = AGG[which(as.character(AGG$Species) %in% tree$tip.label),]
  
  df_vec <- as.character(AGG$Species)
  tree_vec <- tree$tip.label
  
  a <- setdiff(df_vec, tree_vec)
  b <- setdiff(tree_vec, df_vec)
  
  row.names(data) = data$Species
  
  tree2 <- drop.tip(tree, b)
  
  TempData = data[, -1]
  contrasts <- as.data.frame(apply(TempData, 2, pic, tree2))
  names(contrasts) = names(TempData)
  
  ##############
  cor.test(log2(AGG$FemaleMaturityDays),AGG$FrA)
  cor.test(log2(AGG$FemaleMaturityDays),AGG$FrT)
  cor.test(log2(AGG$FemaleMaturityDays),AGG$FrG)
  cor.test(log2(AGG$FemaleMaturityDays),AGG$FrC)
  
  A = lm(log2(AGG$FemaleMaturityDays) ~  scale(AGG$FrA) + scale(AGG$FrT) + scale(AGG$FrC));   summary(A)
  A = lm(log2(AGG$FemaleMaturityDays) ~  scale(AGG$FrT) + scale(AGG$FrC));   summary(A)
  
  cor.test(log2(contrasts$FemaleMaturityDays), contrasts$FrA)
  cor.test(log2(contrasts$FemaleMaturityDays), contrasts$FrT)
  cor.test(log2(contrasts$FemaleMaturityDays), contrasts$FrG)
  cor.test(log2(contrasts$FemaleMaturityDays), contrasts$FrC)
  
  A = lm(log2(contrasts$FemaleMaturityDays) ~  scale(contrasts$FrA) + scale(contrasts$FrT) + scale(contrasts$FrC));   summary(A)
  A = lm(log2(contrasts$FemaleMaturityDays) ~  scale(contrasts$FrA) + scale(contrasts$FrC));   summary(A)

  ###### start from pairwise correlations and go to multiple linear model:
  ## it is opposite s compared to mammals
  par(mfrow=c(2,2),oma = c(0, 0, 2, 0),cex.main = 2, cex.lab = 2)
  plot(log2(AGG$FemaleMaturityDays),AGG$FrA, col = 'gray', ylim = c(0,1), xlab = 'log2(GT)', main = 'FrA'); abline(h =0.4, lt = 2, col = 'red')
  plot(log2(AGG$FemaleMaturityDays),AGG$FrT, col = 'blue', ylim = c(0,1), xlab = 'log2(GT)', main = 'FrT'); abline(h =0.4, lt = 2, col = 'red')
  plot(log2(AGG$FemaleMaturityDays),AGG$FrG, col = 'green', ylim = c(0,1), xlab = 'log2(GT)', main = 'FrG'); abline(h =0.4, lt = 2, col = 'red')
  plot(log2(AGG$FemaleMaturityDays),AGG$FrC, col = 'cyan', ylim = c(0,1), xlab = 'log2(GT)', main = 'FrC'); abline(h =0.4, lt = 2, col = 'red')
  mtext(VecOfTaxaShort[i], outer = TRUE, cex = 1.5)

  ########### question 1: which nucleotides better correlate with GT
  AGG = aggregate(list(SynNucAA$FrA,SynNucAA$FrT,SynNucAA$FrG,SynNucAA$FrC), by = list(SynNucAA$Species,SynNucAA$Maximum.longevity..yrs.), FUN = mean)
  names(AGG) = c('Species','MaxLifeSpan','FrA','FrT','FrG','FrC')
  
  ###### start from pairwise correlations and go to multiple linear model:
  ## it is opposite s compared to mammals
  par(mfrow=c(2,2),oma = c(0, 0, 2, 0),cex.main = 2, cex.lab = 2)
  plot(log2(AGG$MaxLifeSpan),AGG$FrA, col = 'gray', ylim = c(0,1), xlab = 'log2(GT)', main = 'FrA'); abline(h =0.4, lt = 2, col = 'red')
  plot(log2(AGG$MaxLifeSpan),AGG$FrT, col = 'blue', ylim = c(0,1), xlab = 'log2(GT)', main = 'FrT'); abline(h =0.4, lt = 2, col = 'red')
  plot(log2(AGG$MaxLifeSpan),AGG$FrG, col = 'green', ylim = c(0,1), xlab = 'log2(GT)', main = 'FrG'); abline(h =0.4, lt = 2, col = 'red')
  plot(log2(AGG$MaxLifeSpan),AGG$FrC, col = 'cyan', ylim = c(0,1), xlab = 'log2(GT)', main = 'FrC'); abline(h =0.4, lt = 2, col = 'red')
  mtext(VecOfTaxaShort[i], outer = TRUE, cex = 1.5)
  
  
  ########### question 2: which genes better correlate with GT (why T in ATP6,COX3 and ND4 do not correlate with GT and high absolute value - fast replication, no tRNA before them?)
  ## T is negatively and C is positively (T->C)
  #VecOfGenes = c('COX1','COX2','ATP8','ATP6','COX3','ND3','ND4L','ND4','ND5','ND6','CytB','ND1','ND2') # ND1 ND2 ND6
  #length(VecOfGenes)
  #par(mfcol=c(4,13))
  #for (i in 1:length(VecOfGenes))
  #{ # i = 1
  #OneGene = SynNucAA[SynNucAA$Gene == VecOfGenes[i],]
  #main = VecOfGenes[i]
  #plot(log2(OneGene$Female.maturity..days.),OneGene$FrA, col = 'gray', main = main, ylim = c(0,1), xlab = 'log2(FM)') # a bit negative
  #plot(log2(OneGene$Female.maturity..days.),OneGene$FrT, col = 'blue', main = main, ylim = c(0,1), xlab = 'log2(FM)') # a bit negative
  #plot(log2(OneGene$Female.maturity..days.),OneGene$FrG, col = 'green', main = main, ylim = c(0,0.6), xlab = 'log2(FM)') # a bit negative
  #plot(log2(OneGene$Female.maturity..days.),OneGene$FrC, col = 'cyan', main = main, ylim = c(0,0.6), xlab = 'log2(FM)') # a bit negative
  #}
}  
dev.off()

